File: packages/reliai-python/reliai/instrumentation/anthropic.py
from __future__ import annotations

from typing import Any


def is_anthropic_request(url: str) -> bool:
    return "api.anthropic.com" in url.lower()


def build_anthropic_trace(
    request_json: dict[str, Any], response_json: dict[str, Any], latency_ms: int, ok: bool
) -> dict[str, Any] | None:
    model = request_json.get("model")
    if not isinstance(model, str):
        return None
    messages = request_json.get("messages")
    prompt_lines: list[str] = []
    if isinstance(messages, list):
        for message in messages:
            if not isinstance(message, dict):
                continue
            role = message.get("role", "message")
            content = _coerce_blocks(message.get("content"))
            if content:
                prompt_lines.append(f"{role}: {content}")
    usage = response_json.get("usage") if isinstance(response_json, dict) else None
    return {
        "model": model,
        "provider": "anthropic",
        "input_text": "\n".join(prompt_lines) or None,
        "output_text": _coerce_blocks(response_json.get("content")) if isinstance(response_json, dict) else None,
        "latency_ms": latency_ms,
        "prompt_tokens": usage.get("input_tokens") if isinstance(usage, dict) else None,
        "completion_tokens": usage.get("output_tokens") if isinstance(usage, dict) else None,
        "success": ok,
        "metadata": {"auto_instrumented": True, "span_type": "llm"},
    }


def _coerce_blocks(content: Any) -> str:
    if isinstance(content, str):
        return content
    if not isinstance(content, list):
        return ""
    return "\n".join(
        block["text"] for block in content if isinstance(block, dict) and isinstance(block.get("text"), str)
    )

File: packages/reliai-node/src/instrumentation/anthropic.ts
import type { ReliaiTraceEvent } from "../tracing.ts";

function coerceBlocks(content: unknown): string {
  if (typeof content === "string") {
    return content;
  }
  if (!Array.isArray(content)) {
    return "";
  }
  return content
    .map((block) => {
      if (!block || typeof block !== "object") {
        return "";
      }
      if ("text" in block && typeof block.text === "string") {
        return block.text;
      }
      return "";
    })
    .filter(Boolean)
    .join("\n");
}

export function isAnthropicRequest(url: string): boolean {
  return /api\.anthropic\.com/i.test(url);
}

export function buildAnthropicTrace(input: {
  url: string;
  requestBody: unknown;
  responseBody: unknown;
  latencyMs: number;
  ok: boolean;
  framework?: string | null;
}): ReliaiTraceEvent | null {
  const requestBody = input.requestBody;
  if (!requestBody || typeof requestBody !== "object") {
    return null;
  }
  const model = "model" in requestBody && typeof requestBody.model === "string" ? requestBody.model : null;
  if (!model) {
    return null;
  }
  const prompt = Array.isArray((requestBody as { messages?: unknown[] }).messages)
    ? (requestBody as { messages: unknown[] }).messages
        .map((message) => {
          if (!message || typeof message !== "object") {
            return "";
          }
          const role = "role" in message && typeof message.role === "string" ? message.role : "message";
          const content = "content" in message ? coerceBlocks(message.content) : "";
          return content ? `${role}: ${content}` : "";
        })
        .filter(Boolean)
        .join("\n")
    : undefined;
  const responseBody = input.responseBody;
  const usage =
    responseBody && typeof responseBody === "object" && "usage" in responseBody && responseBody.usage
      ? responseBody.usage
      : null;
  return {
    model,
    provider: "anthropic",
    input: prompt,
    output:
      responseBody && typeof responseBody === "object" && "content" in responseBody
        ? coerceBlocks(responseBody.content) || null
        : null,
    latencyMs: input.latencyMs,
    promptTokens: usage && typeof usage === "object" && "input_tokens" in usage ? Number(usage.input_tokens) : undefined,
    completionTokens:
      usage && typeof usage === "object" && "output_tokens" in usage ? Number(usage.output_tokens) : undefined,
    success: input.ok,
    metadata: {
      auto_instrumented: true,
      span_type: "llm",
      framework: input.framework ?? undefined,
      provider_endpoint: input.url,
    },
  };
}
